File: whatsapp_bot.py
import os
import time
from groq import Groq
from dotenv import load_dotenv
from whatsapp_bot import WhatsAppBot

# Load environment variables
load_dotenv()

# Initialize Groq client
client = Groq(
    api_key=os.getenv('GROQ_API_KEY', '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')
)

class AIWhatsAppBot:
    def __init__(self):
        self.phone_number = os.getenv('WHATSAPP_PHONE')
        if not self.phone_number:
            raise ValueError("Please set WHATSAPP_PHONE in your .env file")
        
        # Initialize WhatsApp bot
        self.bot = WhatsAppBot()
        
    def generate_ai_response(self, query):
        """Generate AI response using Groq"""
        try:
            chat_completion = client.chat.completions.create(
                messages=[
                    {
                        "role": "system",
                        "content": "أنت مساعد ذكي يجيب باللغة العربية. قدم إجابات مفيدة ودقيقة ومختصرة."
                    },
                    {
                        "role": "user",
                        "content": query
                    }
                ],
                model="llama3-70b-8192",
                temperature=0.7,
                max_tokens=500,
            )
            return chat_completion.choices[0].message.content
        except Exception as e:
            print(f"Error generating AI response: {e}")
            return "عذراً، حدث خطأ في توليد الرد. الرجاء المحاولة مرة أخرى."

    def process_message(self, message):
        """Process incoming message"""
        if message.startswith('!ai'):
            query = message[4:].strip()
            print(f"Processing query: {query}")
            response = self.generate_ai_response(query)
            print(f"AI Response: {response}")
            return self.bot.send_message(self.phone_number, response)
        return False

    def start(self):
        """Start the bot"""
        print("WhatsApp Bot Starting...")
        
        try:
            # Send welcome message
            welcome_message = "مرحباً! أنا بوت الذكاء الاصطناعي. أرسل '!ai' متبوعاً برسالتك للتفاعل معي."
            self.bot.send_message(self.phone_number, welcome_message)
            
            print("\nInstructions:")
            print("1. Scan the QR code when it appears")
            print("2. Send messages starting with '!ai'")
            print("Example: !ai ما هو أفضل لغة برمجة للمبتدئين؟")
            
            # Start listening for messages
            print("\nBot is running. Press Ctrl+C to stop.")
            self.bot.start()
            
        except KeyboardInterrupt:
            print("\nBot stopped by user.")
        except Exception as e:
            print(f"Error: {e}")

def main():
    bot = AIWhatsAppBot()
    bot.start()

if __name__ == "__main__":
    main()

File: whatsapp_bot.js
const { Client, LocalAuth, MessageMedia } = require('whatsapp-web.js');
const qrcode = require('qrcode-terminal');
const { GoogleGenerativeAI } = require('@google/generative-ai');
const { HfInference } = require('@huggingface/inference');
const fs = require('fs');
const path = require('path');
const https = require('https');
require('dotenv').config();

// Initialize AI clients
const genAI = new GoogleGenerativeAI(process.env.GEMINI_API_KEY);
const hf = new HfInference(process.env.HUGGINGFACE_API_KEY);

// Helper function to download image
async function downloadImage(url, filepath) {
    return new Promise((resolve, reject) => {
        https.get(url, (res) => {
            if (res.statusCode === 200) {
                res.pipe(fs.createWriteStream(filepath))
                   .on('error', reject)
                   .once('close', () => resolve(filepath));
            } else {
                res.resume();
                reject(new Error(`Request Failed With a Status Code: ${res.statusCode}`));
            }
        });
    });
}

// Initialize WhatsApp client
const client = new Client({
    authStrategy: new LocalAuth(),
    puppeteer: {
        args: ['--no-sandbox']
    }
});

// Generate QR code
client.on('qr', (qr) => {
    console.log('Please scan this QR code to connect to WhatsApp Web:');
    qrcode.generate(qr, { small: true });
});

// Ready event
client.on('ready', () => {
    console.log('WhatsApp Bot is ready!');
    console.log('\nInstructions:');
    console.log('1. Send messages starting with !ai');
    console.log('2. Send images with caption starting with !ai');
    console.log('Example: !ai What is in this image?');
});

// Handle incoming messages
client.on('message', async msg => {
    if (msg.body.startsWith('!ai')) {
        const query = msg.body.slice(4).trim();
        console.log(`Processing query: ${query}`);

        try {
            let response;
            
            // Check if message has media
            if (msg.hasMedia) {
                const media = await msg.downloadMedia();
                if (media.mimetype.startsWith('image/')) {
                    // Save image temporarily
                    const tempPath = path.join(__dirname, 'temp_image.jpg');
                    fs.writeFileSync(tempPath, Buffer.from(media.data, 'base64'));

                    try {
                        // Use Hugging Face for image analysis
                        const imageAnalysis = await hf.imageToText({
                            model: 'Salesforce/blip-image-captioning-large',
                            data: fs.readFileSync(tempPath)
                        });

                        // Get text response from Gemini
                        const model = genAI.getGenerativeModel({ model: "gemini-pro" });
                        const prompt = `Image description: ${imageAnalysis.generated_text}\nUser question: ${query}\nPlease provide a detailed response about the image based on this information.`;
                        
                        const result = await model.generateContent(prompt);
                        response = await result.response;

                    } catch (error) {
                        console.error('Error analyzing image:', error);
                        response = { text: () => "Sorry, there was an error analyzing the image. Please try again." };
                    }

                    // Clean up temp file
                    fs.unlinkSync(tempPath);
                } else {
                    response = { text: () => "Sorry, I can only process images at the moment." };
                }
            } else {
                // Text-only query
                const model = genAI.getGenerativeModel({ model: "gemini-pro" });
                const result = await model.generateContent(query);
                response = await result.response;
            }

            const text = response.text();
            console.log(`AI Response: ${text}`);
            await msg.reply(text);
        } catch (error) {
            console.error('Error:', error);
            await msg.reply('Sorry, there was an error generating the response. Please try again.');
        }
    }
});

// Error handling
client.on('auth_failure', () => {
    console.error('Authentication failed');
});

client.on('disconnected', (reason) => {
    console.log('Client disconnected:', reason);
});

// Initialize client
console.log('Starting WhatsApp Bot...');
client.initialize();
